Meter.c

```diff
@@ -266,89 +315,571 @@ static void BarMeterMode_draw(Meter* this, int x, int y, int w) {
+               if (cellsPainted == topCell - 1) {
+                  assert(blanksAtTopCell < 8);
+                  record[offset].cell.details &= (0xFF << blanksAtTopCell);
+               } else if (numCells == 1) {
+                  assert(blanksAtEnd < 8);
+                  record[offset].cell.details &= (0xFF << blanksAtEnd);
+               }
+               numCells--;
+               cellsPainted++;
+            }
+            cellsPainted += numCells;
+
+            bottom = top;
+         }
+
+         if (cellsPainted == topCell) {
+            break;
+         }
+         // Set new bounds and threshold
+         if (cellsPainted > topCell) {
+            paintedHigh = cellsPainted;
+            if (thresholdLow >= thresholdHigh) {
+               if (cellLimitLow >= cellLimitHigh) {
+                  assert(tiedItemLimit >= topCellItem);
+                  tiedItemLimit = numTiedItems - (uint8_t)(cellsPainted - topCell);
+               } else {
+                  assert(cellLimitHigh > cellLimit);
+                  cellLimitHigh = cellLimit;
+               }
+            } else {
+               assert(thresholdLow < threshold);
+               thresholdLow = threshold;
+            }
+         } else {
+            paintedLow = cellsPainted + 1;
+            if (thresholdLow >= thresholdHigh) {
+               assert(cellLimitLow < cellLimitHigh);
+               assert(cellLimitLow < nextCellLimitLow);
+               cellLimitLow = nextCellLimitLow;
+               nextCellLimitHigh = cellLimitHigh;
+            } else {
+               assert(cellLimit >= topCell);
+               assert(thresholdHigh > nextThresholdHigh);
+               thresholdHigh = nextThresholdHigh;
+               nextThresholdLow = thresholdLow;
+            }
+         }
+         threshold = thresholdHigh;
+         if (thresholdLow >= thresholdHigh) {
+            cellLimit = cellLimitLow;
+            if (cellLimitLow < cellLimitHigh && paintedHigh > paintedLow) {
+               cellLimit += ((cellLimitHigh - cellLimitLow) *
+                            (topCell - paintedLow) /
+                            (paintedHigh - paintedLow));
+               if (cellLimit > nextCellLimitHigh) {
+                  cellLimit = nextCellLimitHigh;
+               }
+            }
+         } else {
+            if (paintedHigh > paintedLow) {
+               threshold -= ((thresholdHigh - thresholdLow) *
+                            (topCell - paintedLow) /
+                            (paintedHigh - paintedLow));
+               if (threshold < nextThresholdLow) {
+                  threshold = nextThresholdLow;
+               }
+            }
+         }
+         assert(threshold <= thresholdHigh);
+      }
+
+      scaleExp++;
+   } while ((!(this->total > 0.0)) && (1U << scaleExp) < GRAPH_HEIGHT * 2);
+}
+
+static int GraphMeterMode_getCellColor(const Meter* this, unsigned int recordIndex, unsigned int h, int scaleExp, uint8_t* details) {
```

Comment:
Excessive cyclomatic complexity. Please refactor to make this function more manageable …

Reply:
I've split a subroutine out to `GraphMeterMode_scaleCellDetails`. How is it then?

Reply:
That's a start, but IMHO that `getCellColor` routine still does much more processing than a simple getter should perform. At the moment this routine is still borderline, and by far not the worst offender. The worst offender right now is `GraphMeterMode_computeColors` AFAIC.